### PR TITLE
Machine controller webhook

### LINKED ISSUE #227: Implement admission controller for validation and defaulting

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubermatic/machine-controller#227

**Special notes for your reviewer**:

**Documentation**:

**Release note**:

```release-note
Kubermatic now validates nodes synchronously
```

### CODE REVIEW

File: api/pkg/resources/machinecontroller/deployment.go [hunk 1]
```diff
@@ -16,7 +16,7 @@ import (
 const (
 	name = "machine-controller"
 
-	tag = "v0.9.9"
+	tag = "65cb55a75723f09f436dc2773ecce3054d8f2217"
```
Comment (alvaroaleman) [inline api/pkg/resources/machinecontroller/deployment.go]: I will change this once the test+build for the machine-controller succeeded

File: api/pkg/cluster/client/client.go [hunk 2]
```diff
@@ -109,6 +110,15 @@ func (p *Provider) GetApiextensionsClient(c *kubermaticv1.Cluster) (apiextension
 	return apiextensionsclientset.NewForConfig(config)
 }
 
+// GetAdmissionRegistrationClient returns a client to interact with admissionregistration resources
+func (p *Provider) GetAdmissionRegistrationClient(c *kubermaticv1.Cluster) (admissionregistrationclientset.AdmissionregistrationV1beta1Interface, error) {
```
Comment (mrIncompetent) [inline api/pkg/cluster/client/client.go:114]: More clients! Because more=better

File: api/pkg/controller/cluster/userclusterresources.go [hunk 3]
```diff
@@ -443,6 +452,45 @@ func (cc *Controller) userClusterEnsureCustomResourceDefinitions(c *kubermaticv1
 	return nil
 }
 
+func (cc *Controller) userClusterEnsureMutatingWebhookConfigurations(c *kubermaticv1.Cluster, data *resources.TemplateData) error {
+	client, err := cc.userClusterConnProvider.GetAdmissionRegistrationClient(c)
+	if err != nil {
+		return err
+	}
+
+	mutatingWebhookConfiguration, err := machinecontroller.MutatingwebhookConfiguration(c, data, nil)
```
Comment (mrIncompetent) [inline api/pkg/controller/cluster/userclusterresources.go]: Can we stick to the `loop` approach ? 

Sure - Its premature optimization, but it fits better in our existing codebase & makes it more easy to read

File: api/pkg/controller/cluster/userclusterresources.go [hunk 4]
```diff
@@ -443,6 +452,45 @@ func (cc *Controller) userClusterEnsureCustomResourceDefinitions(c *kubermaticv1
...
+	existing, err := client.MutatingWebhookConfigurations().Get(mutatingWebhookConfiguration.Name, metav1.GetOptions{})
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+
+		if _, err = client.MutatingWebhookConfigurations().Create(mutatingWebhookConfiguration); err != nil {
+			return fmt.Errorf("failed to create MutatingWebhookConfiguration: %v", err)
+		}
+		glog.V(4).Infof("Created MutatingWebhookConfiguration %s", mutatingWebhookConfiguration.Name)
```
Comment (mrIncompetent) [inline api/pkg/controller/cluster/userclusterresources.go]: When logging, add the cluster name + the fact that we created it inside the user cluster.
```go
glog.V(4).Infof("Created MutatingWebhookConfiguration %s inside user cluster %s", mutatingWebhookConfiguration.Name, c.Name)
```

File: api/pkg/controller/cluster/userclusterresources.go [hunk 5]
```diff
@@ -443,6 +452,45 @@ func (cc *Controller) userClusterEnsureCustomResourceDefinitions(c *kubermaticv1
...
+	}
+
+	mutatingWebhookConfiguration, err = machinecontroller.MutatingwebhookConfiguration(c, data, existing.DeepCopy())
+	if err != nil {
+		return fmt.Errorf("failed to build MutatingWebhookConfigurations%v", err)
```
Comment (mrIncompetent) [inline api/pkg/controller/cluster/userclusterresources.go]:
```go
return fmt.Errorf("failed to build MutatingWebhookConfiguration: %v", err)
```

File: api/pkg/resources/machinecontroller/deployment.go (hunk 1 above)
Comment (mrIncompetent) [inline api/pkg/resources/machinecontroller/deployment.go]: Why are we using a hash instead of a version?

File: api/pkg/resources/machinecontroller/deployment.go [hunk 6]
```diff
@@ -131,17 +131,15 @@ func Deployment(data resources.DeploymentDataProvider, existing *appsv1.Deployme
...
-		{
-			Name: resources.MachineControllerKubeconfigSecretName,
-			VolumeSource: corev1.VolumeSource{
-				Secret: &corev1.SecretVolumeSource{
-					SecretName: resources.MachineControllerKubeconfigSecretName,
-					// We have to make the secret readable for all for now because owner/group cannot be changed.
-					// ( upstream proposal: https://github.com/kubernetes/kubernetes/pull/28733 )
-					DefaultMode: resources.Int32(resources.DefaultAllReadOnlyMode),
-				},
+func getKubeconfigVolume() corev1.Volume {
```
Comment (mrIncompetent) [inline api/pkg/resources/machinecontroller/deployment.go:134]: Why was this changed?

File: api/pkg/resources/machinecontroller/deployment.go (hunk 1 above)
Comment (mrIncompetent) [reply api/pkg/resources/machinecontroller/deployment.go]: Just seen your comment above

**Comment by mrIncompetent** [comment]
@alvaroaleman have you checked that we do not trigger any defaulting on the seed APIserver with the new resources?

So we don't trigger an update on each sync

File: api/pkg/resources/machinecontroller/deployment.go (hunk 1 above)
Comment (alvaroaleman) [reply api/pkg/resources/machinecontroller/deployment.go]: As commented above, because the build did not suceed yet :disappointed:

File: api/pkg/resources/machinecontroller/deployment.go (hunk 6 above)
Comment (alvaroaleman) [reply api/pkg/resources/machinecontroller/deployment.go:134]: Because the webhook needs an additional cert, hence I changed the signature to not return a slice bit a single volume

**Comment by alvaroaleman** [comment]
> @alvaroaleman have you checked that we do not trigger any defaulting on the seed APIserver with the new resources?
So we don't trigger an update on each sync

Yes. I've also created #2342 as an issue to have a test for this

**Comment by alvaroaleman** [comment]
@mrIncompetent All comments adressed, PTAL

**Review by mrIncompetent — APPROVED** [review]
